cpp/src/arrow/dataset/scanner_test.cc

```diff
@@ -1445,9 +1446,13 @@ DatasetAndBatches MakeBasicDataset() {
           equal(field_ref("c"), literal(23)),
           equal(field_ref("c"), literal(47)),
       },
-      [](compute::ExecBatch* batch, const RecordBatch&) {
+      [](compute::ExecBatch* batch, const RecordBatch&, size_t guarantee_index) {
         // a placeholder will be inserted for partition field "c"
-        batch->values.emplace_back(std::make_shared<Int32Scalar>());
+        if (guarantee_index == 0) {
+          batch->values.emplace_back(std::make_shared<Int32Scalar>(23));
+        } else {
+          batch->values.emplace_back(std::make_shared<Int32Scalar>(47));
+        }
```

Comment:
This seems like a spot-fix.  Perhaps a more general fix would be to fix the `DatasetAndBatchesFromJSON` function?

Reply:
I address this in .

I essentially change the invocation of the lambda to:
```cpp
if (not guarantees.empty()) {
  auto extract_result = ExtractKnownFieldValues(guarantees[frag_ndx]);
  ARROW_WARN_NOT_OK(extract_result.status(), "ExtractKnownFieldValues failed");
  for (const auto& known_field : extract_result->map) {
    batches.back().values.emplace_back(known_field.second);
  }
}
```

ooh, but I just realized I used `not` instead of `!`. I'm not sure if there's a preference

Reply:
oddly enough, I can't mark this resolved.